### PR TITLE
[Prim] Add prod backward composite rule

### PR DESCRIPTION
**Pull request by rainyfly**
### PR types

Others
### PR changes

OPs
### Describe

1. Add composite rules for prod backward.

### CODE REVIEW

**Comment by rainyfly** [comment]
TestProd8DOp已禁用，因为unqueeze和reshape都不支持8D

File: paddle/fluid/prim/api/composite_backward/composite_backward_api.h [hunk 1]
```diff
@@ -887,6 +887,62 @@ void gather_nd_grad(const Tensor& x,
...
+          for (int64_t i = 0; i < axis_size; i++) {
+            if (axis[i] < 0) {
+              axis_[i] = axis[i] + x_dim_size;
+            }
+          }
+        }
+        auto reshape_dim = std::vector<int64_t>(x_dim);
+        for (auto i : axis_) {
+          reshape_dim[i] = 1;
+        }
```
Comment (xiaoguoguo626807) [inline paddle/fluid/prim/api/composite_backward/composite_backward_api.h]: use unsqueezeapi  replace

File: python/paddle/fluid/tests/unittests/test_reduce_op.py [hunk 2]
```diff
@@ -365,16 +367,20 @@ def test_check_output(self):
...
 
 class TestProdOp_ZeroDim(OpTest):
     def setUp(self):
         self.python_api = paddle.prod
         self.op_type = "reduce_prod"
+        self.prim_op_type = "prim"
         self.inputs = {'X': np.random.random([]).astype("float64")}
         self.outputs = {'Out': self.inputs['X'].prod()}
         self.attrs = {'dim': [], 'reduce_all': True}
+        # reduce doesn't support float64 in cinn.
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/test_reduce_op.py]: will it raise error when test fp64 with other test(not 0D tensor) in cinn?

File: python/paddle/fluid/tests/unittests/test_reduce_op.py [hunk 3]
```diff
@@ -405,13 +412,14 @@ def test_check_output(self):
...
     def test_check_grad(self):
-        self.check_grad(['X'], 'Out', check_eager=True)
+        self.check_grad(['X'], 'Out', check_eager=True, check_prim=True)
 
 
 class TestProd8DOp(OpTest):
     def setUp(self):
         self.op_type = "reduce_prod"
         self.python_api = raw_reduce_prod
+        self.prim_op_type = "prim"
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/test_reduce_op.py]: if don't test prim delete this

File: python/paddle/fluid/tests/unittests/test_reduce_op.py [hunk 4]
```diff
@@ -387,6 +393,7 @@ class TestProd6DOp(OpTest):
...
         self.init_data_type()
         self.inputs = {
             'X': np.random.random((5, 6, 2, 3, 4, 2)).astype(self.data_type)
```
Comment (xiaoguoguo626807) [inline python/paddle/fluid/tests/unittests/test_reduce_op.py:462]: confirm all dtypes  has been tested (FP16, FP32,Fp64)

File: paddle/fluid/prim/api/composite_backward/composite_backward_api.h (hunk 1 above)
Comment (rainyfly) [reply paddle/fluid/prim/api/composite_backward/composite_backward_api.h]: done

File: python/paddle/fluid/tests/unittests/test_reduce_op.py (hunk 2 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_reduce_op.py]: fp64 added in test

File: python/paddle/fluid/tests/unittests/test_reduce_op.py (hunk 3 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_reduce_op.py]: done

File: python/paddle/fluid/tests/unittests/test_reduce_op.py (hunk 4 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_reduce_op.py:462]: FP16 is not supported, Fp64 has been added

**Review by xiaoguoguo626807 — COMMENTED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/test_reduce_op.py (hunk 4 above)
Comment (rainyfly) [reply python/paddle/fluid/tests/unittests/test_reduce_op.py:462]: 591602a47371af7ac4432f53a88fe2b2

**Review by Charles-hit — APPROVED** [review]

**Review by xiaoguoguo626807 — APPROVED** [review]
LGTM

**Pull request merged**